verification/io/io-spec.gobra

```diff
@@ -38,75 +38,191 @@ type Place int
+pure func dp3s_iospec_bio3s_enter_guard(s IO_dp3s_state_local, t Place, v IO_val) bool {
+	return some(v.IO_Internal_val1_2) in domain(s.ibuf) ==>
+		(let ibuf_set := s.ibuf[some(v.IO_Internal_val1_2)] in (v.IO_Internal_val1_1 in ibuf_set)) &&
+		(exists currseg IO_seg3, traversedseg IO_seg3 ::
+		(exists hf1 IO_HF, fut seq[IO_HF] ::
+			dp2_enter_guard(
+				v.IO_Internal_val1_1,
+				currseg,
+				traversedseg,
+				asid(),
+				hf1,
+				v.IO_Internal_val1_2,
+				fut) &&
+				dp3s_forward(
+					IO_pkt2(
+						IO_Packet2{
+							traversedseg,
+							v.IO_Internal_val1_1.LeftSeg,
+							v.IO_Internal_val1_1.MidSeg,
+							v.IO_Internal_val1_1.RightSeg}),
+					v.IO_Internal_val1_3,
+					v.IO_Internal_val1_4)))
+}
+
+pred dp3s_iospec_bio3s_enter(s IO_dp3s_state_local, t Place) {
+	forall v IO_val ::
+	(v.isIO_Internal_val1 ==>
+		(dp3s_iospec_bio3s_enter_guard(s, t, v) ==>
+			(CBio_IN_bio3s_enter(t, v) &&
+				dp3s_iospec_ordered(
+					dp3s_add_obuf(s, v.IO_Internal_val1_4, v.IO_Internal_val1_3),
+					CBio_IN_bio3s_enter_T(s, t, v)))))
+}
+
+pred CBio_IN_bio3s_xover_up2down(t Place, v IO_val)
+
+ghost
+requires CBio_IN_bio3s_xover_up2down(t, v)
+pure func dp3s_iospec_bio3s_xover_up2down_T(s IO_dp3s_state_local, t Place, v IO_val) Place
+
+
+// This corresponds to the condition of the if statement in the io-spec case for xover_up2down
+ghost
+decreases
+pure func dp3s_iospec_bio3s_xover_up2down_guard(s IO_dp3s_state_local, t Place, v IO_val) bool {
+	return (exists currseg IO_seg3 :: !currseg.ConsDir &&
+		(exists nextseg IO_seg3 :: nextseg.ConsDir &&
+			(exists traversedseg IO_seg3, intermediatepkt IO_pkt3, hf1 IO_HF, hf2 IO_HF ::
+			xover_up2down2_link_type(asid(), hf1, hf2) &&
+				(exists nextfut seq[IO_HF] ::
+				dp3s_xover_common(
+					s,
+					v.IO_Internal_val1_1,
+					currseg,
+					nextseg,
+					traversedseg,
+					intermediatepkt,
+					hf1,
+					hf2,
+					nextfut,
+					v.IO_Internal_val1_2,
+					v.IO_Internal_val1_3,
+					v.IO_Internal_val1_4,)))))
+}
+
+pred dp3s_iospec_bio3s_xover_up2down(s IO_dp3s_state_local, t Place) {
+	forall v IO_val ::
+	(v.isIO_Internal_val1 ==>
+		(dp3s_iospec_bio3s_xover_up2down_guard(s, t, v) ==>
+			(CBio_IN_bio3s_xover_up2down(t, v) &&
+				dp3s_iospec_ordered(
+					dp3s_add_obuf(s, v.IO_Internal_val1_4, v.IO_Internal_val1_3),
+					dp3s_iospec_bio3s_xover_up2down_T(s, t, v)))))
+}
+
+pred CBio_IN_bio3s_xover_core(t Place, v IO_val)
+
+ghost
+requires CBio_IN_bio3s_xover_core(t, v)
+pure func dp3s_iospec_bio3s_xover_core_T(s IO_dp3s_state_local, t Place, v IO_val) Place
+
+// This corresponds to the condition of the if statement in the io-spec case for xover_core
+ghost
+decreases
+pure func dp3s_iospec_bio3s_xover_core_guard(s IO_dp3s_state_local, t Place, v IO_val) bool {
+	return (asid() in core_as_set() &&
+		(exists currseg IO_seg3, nextseg IO_seg3 ::
+		currseg.ConsDir == nextseg.ConsDir &&
+			(exists traversedseg IO_seg3, intermediatepkt IO_pkt3, hf1 IO_HF, hf2 IO_HF ::
```

Comment:
can you please check if the following existential quantifiers in this function and those that follow can be removed in a similar way to the one proposed above?